services/rollout-service/pkg/datadogdora/datadogdora.go

```diff
@@ -0,0 +1,205 @@
+	// used to report dora metrics
+	doraAPI *datadogV2.DORAMetricsApi
+}
+
+func (s *Subscriber) Subscribe(ctx context.Context, b *service.Broadcast) error {
+	if s.state == nil {
+		s.state = map[service.Key]*service.BroadcastEvent{}
+	}
+	for {
+		err := s.subscribeOnce(ctx, b)
+		select {
+		case <-ctx.Done():
+			return err
+		default:
+		}
+	}
+}
+
+func (s *Subscriber) subscribeOnce(ctx context.Context, b *service.Broadcast) error {
+	event, ch, unsub := b.Start()
+	defer unsub()
+	for _, ev := range event {
+		if ev.IsProduction != nil && *ev.IsProduction {
+			s.state[ev.Key] = ev
+		}
+	}
+	s.ready()
+	for {
+		select {
+		case <-ctx.Done():
+			return s.group.Wait()
+		case ev, ok := <-ch:
+			if !ok {
+				return s.group.Wait()
+			}
+			if ev.IsProduction == nil || !*ev.IsProduction {
+				continue
+			}
+			if s.maxAge != 0 &&
+				ev.ArgocdVersion != nil &&
+				ev.ArgocdVersion.DeployedAt.Add(s.maxAge).Before(s.now()) {
+				continue
+			}
+			if shouldNotify(s.state[ev.Key], ev) {
+				s.group.Go(s.notify(ctx, ev))
+			}
+			s.state[ev.Key] = ev
+		}
+	}
+}
+
+func shouldNotify(old *service.BroadcastEvent, nu *service.BroadcastEvent) bool {
+	// check for fields that must be present to generate the request
+	if nu.ArgocdVersion == nil || nu.IsProduction == nil || nu.ArgocdVersion.SourceCommitId == "" {
+		return false
+	}
+	if old == nil || old.ArgocdVersion == nil || old.IsProduction == nil {
+		return true
+	}
+	if old.ArgocdVersion.SourceCommitId != nu.ArgocdVersion.SourceCommitId || old.ArgocdVersion.DeployedAt != nu.ArgocdVersion.DeployedAt {
+		return true
+	}
+	return false
+}
+func (s *Subscriber) notify(ctx context.Context, ev *service.BroadcastEvent) func() error {
+
+	return func() error {
+		span, _ := tracer.StartSpanFromContext(ctx, "datadogdora.notify")
+		defer span.Finish()
+		span.SetTag("datadogAPI.url", s.url)
+		span.SetTag("environment", ev.Environment)
+		span.SetTag("application", ev.Application)
+		// nolint
+		body := datadogV2.DORADeploymentRequest{
+			AdditionalProperties: nil,
+			UnparsedObject:       nil,
+			Data: datadogV2.DORADeploymentRequestData{
+				AdditionalProperties: nil,
+				UnparsedObject:       nil,
+				Attributes: datadogV2.DORADeploymentRequestAttributes{
+					Env:                  &ev.Environment,
+					Id:                   &ev.KuberpultVersion.SourceCommitId,
+					AdditionalProperties: nil,
+					UnparsedObject:       nil,
+					FinishedAt:           ev.ArgocdVersion.DeployedAt.UnixNano(),
+					Git: &datadogV2.DORAGitInfo{
+						AdditionalProperties: nil,
+						UnparsedObject:       nil,
+						CommitSha:            ev.ArgocdVersion.SourceCommitId,
+						RepositoryUrl:        s.RepositoryUrl,
+					},
+					Service: ev.Application,
+					// TODO(BJ) get the time the sync was triggered?
+					StartedAt: ev.ArgocdVersion.DeployedAt.UnixNano() - 1,
+					Version:   datadog.PtrString("v1.12.07"),
+				},
+			},
+		}
+		ctx = context.WithValue(
+			context.Background(),
```

Comment:
why the background context? That sounds wrong.
I think you want the context returned from `StartSpanFromContext` here?

Reply:
I would also say that